genhost

```diff
@@ -18,50 +18,71 @@ rand() {
-  printf 'Generate unused hostnames by randomly picking from a word list\n'
+  printf 'Usage:\n'
+  printf '  genhost +<integer>\n    Generate hostnames by randomly picking unused names from the wordlist\n'
+  printf '  genhost -<hostname>\n    Return a decommissioned hostname to the available pool\n'
+  printf '  genhost ?\n    Print a list of all hostnames currently in use\n'
   exit 1
 else
-  if [[ $1 == *[!0-9]* ]]; then
-    printf 'genhost: "%d" is not a positive integer\n' "$1" >&2
-    exit 1
-  fi
-fi
+  if [[ $1 == +[0-9]* ]]; then
+    # read the word list into memory, removing commented words;
+    # then remove the trailing empty element, if any:
+    #    http://mywiki.wooledge.org/BashFAQ/005#Handling_newlines_.28or_lack_thereof.29_at_the_end_of_a_file
+    unset lines n
+    while IFS= read -r 'lines[n++]'; do :; done < <(sed '/^#/d' "$wordlist")
+      [[ ${lines[n-1]} ]] || unset -v 'lines[--n]'
 
-# read the word list into memory, removing commented words;
-# then remove the trailing empty element, if any:
-#    http://mywiki.wooledge.org/BashFAQ/005#Handling_newlines_.28or_lack_thereof.29_at_the_end_of_a_file
-unset lines n
-while IFS= read -r 'lines[n++]'; do :; done < <(sed '/^#/d' "$wordlist")
-[[ ${lines[n-1]} ]] || unset -v 'lines[--n]'
+    # output the random hostname(s)
+    for i in $(seq $1); do
+      if [[ $n -eq 0 ]]; then
+        printf 'genhost: no more available hostnames left in file %s\n' "$wordlist" >&2
+        exit 1
+      fi
 
-# output the random hostname(s)
-for i in $(seq $1); do
-  if [[ $n -eq 0 ]]; then
-    printf 'genhost: no more available hostnames left in file %s\n' "$wordlist" >&2
-    exit 1
-  fi
+      # pick a random available word
+      rand $n
+      printf '%s.%s\n' "${lines[r]}" "$domain"
 
-  # pick a random available word
-  rand $n
-  printf '%s.%s\n' "${lines[r]}" "$domain"
+      # mark the word as used in the word list
+      case $OSTYPE in
+        darwin*)
+          sed -i '' "s/^${lines[r]}\$/#${lines[r]}/" "$wordlist"
+          ;;
+        linux*)
+          sed -i "s/^${lines[r]}\$/#${lines[r]}/" "$wordlist"
+          ;;
+        msys*)
+          sed -i "s/^${lines[r]}\$/#${lines[r]}/" "$wordlist"
+          ;;
+      esac
 
-  # mark the word as used in the word list
-  case $OSTYPE in
-    darwin*)
-      sed -i '' "s/^${lines[r]}\$/#${lines[r]}/" "$wordlist"
-      ;;
-    linux*)
-      sed -i "s/^${lines[r]}\$/#${lines[r]}/" "$wordlist"
-      ;;
-    msys*)
-      sed -i "s/^${lines[r]}\$/#${lines[r]}/" "$wordlist"
-      ;;
-  esac
-
-  # remove the used word from the array to prevent duplicates
-  lines=(${lines[@]:0:$r} ${lines[@]:$(( $r + 1 ))})
-  n=$(( n - 1 ))
-done
+      # remove the used word from the array to prevent duplicates
+      lines=(${lines[@]:0:$r} ${lines[@]:$(( $r + 1 ))})
+      n=$(( n - 1 ))
+    done
+    exit 0 
+  elif [[ $1 == -[A-Za-z]* ]]; then
+    # strip the leading dash
+    token=$(echo $1 | tail -c +2)
```

Comment:
For stripping the leading dash, rather than shelling out and launching additional processes, I'd recommend relying on Bash's [parameter expansion](http://wiki.bash-hackers.org/syntax/pe#substring_removal) capabilities to do the substring removal:

    token=${1#-}